dm/worker/server.go

```diff
@@ -75,23 +77,41 @@ func (s *Server) Start() error {
 		return terror.ErrWorkerStartService.Delegate(err)
 	}
 
-	s.worker, err = NewWorker(s.cfg)
-	if err != nil {
-		return err
-	}
-
-	s.wg.Add(2)
-	go func() {
-		defer s.wg.Done()
-		// start running worker to handle requests
-		s.worker.Start()
-	}()
-
+	s.worker = nil
+	s.ctx, s.cancel = context.WithCancel(context.Background())
+	s.wg.Add(1)
 	go func() {
 		defer s.wg.Done()
 		// worker keepalive with master
-		// FIXME: use global context
-		s.KeepAlive(s.worker.ctx)
+		// If worker loses connect from master, it would stop all task and try to connect master again.
+		shouldExit := false
+		shouldStop := false
+		for !shouldExit {
+			shouldExit, err = s.KeepAlive(s.ctx)
+			if err != nil || !shouldExit {
+				if shouldStop {
+					s.Lock()
+					if s.worker != nil {
+						s.worker.Close()
+						s.worker = nil
+					}
+					s.Unlock()
+					shouldStop = false
+				} else {
+					// Try to connect master again before stop worker
```

Comment:
Are `TODO` for this line and L110?

Reply:
Why? This gorutine  will keep trying KeepAlive util it receives a signal which set 'shouldExit' true.